python/ray/rllib/optimizers/multi_gpu.py

```diff
@@ -2,8 +2,103 @@
 from __future__ import division
 from __future__ import print_function
 
+import numpy as np
+import tensorflow as tf
+
+import ray
+from ray.rllib.optimizers.evaluator import TFMultiGPUSupport
 from ray.rllib.optimizers.optimizer import Optimizer
+from ray.rllib.optimizers.sample_batch import SampleBatch
+from ray.rllib.parallel import LocalSyncParallelOptimizer
+from ray.rllib.utils.timer import TimerStat
 
 
 class LocalMultiGPUOptimizer(Optimizer):
-    pass  # TODO(ekl)
+    """A synchronous optimizer that uses multiple local GPUs.
+
+    Samples are pulled synchronously from multiple remote evaluators,
+    concatenated, and then split across the memory of multiple local GPUs.
+    A number of SGD passes are then taken over the in-memory data. For more
+    details, see `ray.rllib.parallel.LocalSyncParallelOptimizer`.
+
+    This optimizer is Tensorflow-specific and require evaluators to implement
+    the TFMultiGPUSupport API.
+    """
+
+    def _init(self):
+        assert isinstance(self.local_evaluator, TFMultiGPUSupport)
+        self.batch_size = self.config.get("sgd_batch_size", 128)
+        gpu_ids = ray.get_gpu_ids()
+        if not gpu_ids:
+            self.devices = ["/cpu:0"]
+        else:
+            self.devices = ["/gpu:{}".format(i) for i in range(len(gpu_ids))]
+        assert self.batch_size > len(self.devices), "batch size too small"
+        self.per_device_batch_size = self.batch_size // len(self.devices)
+        self.sample_timer = TimerStat()
+        self.load_timer = TimerStat()
+        self.grad_timer = TimerStat()
+        self.update_weights_timer = TimerStat()
+
+        print("LocalMultiGPUOptimizer devices", self.devices)
+        print("LocalMultiGPUOptimizer batch size", self.batch_size)
+
+        # List of (feature name, feature placeholder) tuples
+        self.loss_inputs = self.local_evaluator.tf_loss_inputs()
+
+        # per-GPU graph copies created below must share vars with the policy
+        tf.get_variable_scope().reuse_variables()
+
+        self.par_opt = LocalSyncParallelOptimizer(
+            tf.train.AdamOptimizer(self.config.get("sgd_stepsize", 5e-5)),
+            self.devices,
+            [ph for _, ph in self.loss_inputs],
+            self.per_device_batch_size,
+            lambda *ph: self.local_evaluator.build_tf_loss(ph),
+            self.config.get("logdir", "/tmp/ray"))
```

Comment:
why this directory? didn't logging directory change per #1297

Reply:
Changed it to os.getcwd()